File: trading-ui/types/paper-desk.ts
export type PaperAgent = "macro" | "research" | "risk";
export type PaperSessionStatus = "open" | "allocated" | "resolved";
export type PaperAssetType = "stock" | "crypto" | "etf";
export type PaperSourceQuality = "market_data" | "prediction_cache" | "fallback" | "model_inference";

export interface PaperDeskAsset {
  symbol: string;
  display_name: string;
  asset_type: PaperAssetType;
  current_price: number;
  price_source: string;
}

export interface PaperDeskBrief {
  agent: PaperAgent;
  stance: string;
  thesis: string;
  confidence: number;
  evidence: string[];
  invalidation: string;
  source_quality: PaperSourceQuality;
  allocation_guardrails: string;
  generated_at: string;
}

export interface PaperDeskAllocation {
  id: string;
  session_id: string;
  anonymous_id: string;
  allocations: Record<string, number>;
  submitted_at: string;
}

export interface PaperDeskResult {
  id: string;
  allocation_id: string;
  session_id: string;
  returns: Record<string, number>;
  portfolio_return: number;
  benchmark_symbol: string;
  benchmark_return: number;
  alpha: number;
  drawdown: number;
  price_source: string;
  resolved_at: string;
}

export interface PaperDeskSession {
  session_id: string;
  trading_date: string;
  status: PaperSessionStatus;
  paper_only: boolean;
  disclaimer: string;
  starting_cash: number;
  benchmark_symbol: string;
  assets: PaperDeskAsset[];
  briefs: PaperDeskBrief[];
  allocation: PaperDeskAllocation | null;
  result: PaperDeskResult | null;
  generated_at: string;
}

export interface PaperDeskHistoryResponse {
  anonymous_id: string;
  sessions: PaperDeskSession[];
}

export interface PaperDeskAllocationPayload {
  session_id: string;
  anonymous_id: string;
  allocations: Record<string, number>;
}

export interface PaperDeskResolvePayload {
  session_id: string;
  anonymous_id: string;
}

File: ai-service/models/paper_desk.py
"""
Pydantic models for the Paper Desk paper-validation loop.
"""

from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

PaperAgent = Literal["macro", "research", "risk"]
PaperSessionStatus = Literal["open", "allocated", "resolved"]
PaperAssetType = Literal["stock", "crypto", "etf"]

SUPPORTED_PAPER_SYMBOLS = {"NVDA", "AMD", "TSLA", "BTCUSDT", "SPY"}


class PaperDeskAsset(BaseModel):
    symbol: str
    display_name: str
    asset_type: PaperAssetType
    current_price: float = Field(..., gt=0)
    price_source: str


class PaperDeskBrief(BaseModel):
    agent: PaperAgent
    stance: str
    thesis: str
    confidence: float = Field(..., ge=0, le=1)
    evidence: List[str] = Field(default_factory=list)
    invalidation: str
    source_quality: Literal["market_data", "prediction_cache", "fallback", "model_inference"]
    allocation_guardrails: str
    generated_at: datetime


class PaperDeskAllocationRequest(BaseModel):
    session_id: str
    anonymous_id: str = Field(..., min_length=8, max_length=128)
    allocations: Dict[str, float]

    @field_validator("allocations")
    @classmethod
    def validate_allocations(cls, allocations: Dict[str, float]) -> Dict[str, float]:
        normalized = {symbol.upper(): float(weight) for symbol, weight in allocations.items()}
        if not normalized:
            raise ValueError("At least one allocation is required")

        unsupported = sorted(set(normalized) - SUPPORTED_PAPER_SYMBOLS)
        if unsupported:
            raise ValueError(f"Unsupported paper symbols: {', '.join(unsupported)}")

        for symbol, weight in normalized.items():
            if weight < 0 or weight > 100:
                raise ValueError(f"Allocation for {symbol} must be between 0 and 100")

        total = sum(normalized.values())
        if abs(total - 100.0) > 0.01:
            raise ValueError(f"Allocations must total 100, got {total:.2f}")

        return normalized


class PaperDeskResolveRequest(BaseModel):
    session_id: str
    anonymous_id: str = Field(..., min_length=8, max_length=128)


class PaperDeskAllocation(BaseModel):
    id: str
    session_id: str
    anonymous_id: str
    allocations: Dict[str, float]
    submitted_at: datetime


class PaperDeskResult(BaseModel):
    id: str
    allocation_id: str
    session_id: str
    returns: Dict[str, float]
    portfolio_return: float
    benchmark_symbol: str
    benchmark_return: float
    alpha: float
    drawdown: float
    price_source: str
    resolved_at: datetime


class PaperDeskSession(BaseModel):
    session_id: str
    trading_date: str
    status: PaperSessionStatus
    paper_only: bool = True
    disclaimer: str = "Paper-only research workflow. No real-money orders are placed."
    starting_cash: int = 100000
    benchmark_symbol: str = "SPY"
    assets: List[PaperDeskAsset]
    briefs: List[PaperDeskBrief]
    allocation: Optional[PaperDeskAllocation] = None
    result: Optional[PaperDeskResult] = None
    generated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaperDeskHistoryResponse(BaseModel):
    anonymous_id: str
    sessions: List[PaperDeskSession]
